src/common/utils.ts

```diff
@@ -995,3 +995,15 @@ export class TernarySearchTree<K, V> {
 		}
 	}
 }
+
+export async function stringReplaceAsync(str: string, regex: RegExp, asyncFn: (substring: string, ...args: any[]) => Promise<string>): Promise<string> {
+	const promises: Promise<string>[] = [];
+	str.replace(regex, (match, ...args) => {
```

Comment:
It'd be a little faster for large input to keep track of the offsets in the match and replace them without running the regex again.